pkg/datagatherer/istio/istio.go

```diff
@@ -0,0 +1,156 @@
+}
+
+// NewDataGatherer creates a new DataGatherer for a cluster.
+func (c *Config) NewDataGatherer(ctx context.Context) (datagatherer.DataGatherer, error) {
+	if err := c.validate(); err != nil {
+		return nil, err
+	}
+
+	// If the Istio namespace is not set then default it to 'istio-system'.
+	istioNamespace := c.IstioNamespace
+	if istioNamespace == "" {
+		istioNamespace = "istio-system"
+	}
+
+	// If a list of resources is not set then use the data gatherer default list.
+	gvrs := c.Resources
+	if len(gvrs) == 0 {
+		gvrs = defaultGVRs
+	}
+
+	// Create a list of Kubernetes dynamic data gatherers to fetch all the required resources for Istio analysis.
+	var dynamics []datagatherer.DataGatherer
+	for _, gvr := range gvrs {
+		configDynamic := k8s.ConfigDynamic{
+			KubeConfigPath:       c.KubeConfigPath,
+			GroupVersionResource: gvr,
+			ExcludeNamespaces:    c.ExcludeNamespaces,
+			IncludeNamespaces:    c.IncludeNamespaces,
+		}
+		dynamic, err := configDynamic.NewDataGatherer(ctx)
+		if err != nil {
+			// TODO: Might be better to collect errors here rather than just give up straight away.
+			return nil, err
+		}
+		dynamics = append(dynamics, dynamic)
+	}
+
+	// Create an Istio SourceAnalyzer.
+	sa := local.NewSourceAnalyzer(istioSchema.MustGet(), analyzers.AllCombined(),
+		"", resource.Namespace(istioNamespace), nil, true, 30*time.Second)
+
+	return &DataGatherer{
+		sa:       sa,
+		dynamics: dynamics,
+	}, nil
+}
+
+// Fetch retrieves resources from the K8s API and runs Istio analysis.
+func (g *DataGatherer) Fetch() (interface{}, error) {
+
+	// Fetch resources from all data gatherers and accumulate in an Unstructured slice.
+	var allResources []unstructured.Unstructured
+	for _, dynamic := range g.dynamics {
+		rawResources, err := dynamic.Fetch()
+		if err != nil {
+			// Data gatherers will error if the resource kind they are trying to fetch is not present in the cluster.
+			// This could be because the cluster does not yet have Istio installed. However we should still run the
+			// analysis on the resources that are available as it is useful for pre-checking a cluster for future Istio
+			// installations.
+			if cfgErr, ok := err.(*dgerror.ConfigError); ok {
+				if cfgErr.Err == "the server could not find the requested resource" {
+					continue
+				}
+			}
+			return nil, err
+		}
+		resources, ok := rawResources.(*unstructured.UnstructuredList)
+		if !ok {
+			return nil, fmt.Errorf("failed to parse resources loaded from DataGatherer")
+		}
+		allResources = append(allResources, resources.Items...)
+	}
+
+	// Convert the slice of Unstructured resources into a YAML document string.
+	allResourcesYAMLString := ""
+	for _, resource := range allResources {
+		// Marshall the unstructured resource to a YAML string.
+		resourceYAML, err := yaml.Marshal(resource.Object)
+		if err != nil {
+			return nil, err
+		}
+		allResourcesYAMLString = string(resourceYAML) + "---\n" + allResourcesYAMLString
```

Comment:
Might be easier to append to a slice and then strings.Join(slice, "---\n") at the end.